bin/cheribsdtest/cheribsdtest_ipc_tagstripped.c

```diff
@@ -0,0 +1,300 @@
+		cheribsdtest_failure_err("socketpair(PF_LOCAL, SOCK_STREAM)");
+}
+
+static void
+ipc_socket_tcp_stream_create(int *fds)
+{
+	struct sockaddr_in sin;
+	int accept_sock, conn_sock, listen_sock;
+	socklen_t socklen;
+
+	listen_sock = socket(PF_INET, SOCK_STREAM, 0);
+	if (listen_sock < 0)
+		cheribsdtest_failure_err("socket(PF_INET, SOCK_STREAM, 0)");
+
+	/* Bind an automatically allocated port; query and connect to it. */
+	bzero(&sin, sizeof(sin));
+	sin.sin_len = sizeof(sin);
+	sin.sin_family = AF_INET;
+	sin.sin_port = 0;
+	sin.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+	if (bind(listen_sock, (struct sockaddr *)&sin, sizeof(sin)) < 0)
+		cheribsdtest_failure_err("bind()");
+	if (listen(listen_sock, -1) < 0)
+		cheribsdtest_failure_err("listen()");
+	socklen = sizeof(sin);
+	if (getsockname(listen_sock, (struct sockaddr *)&sin, &socklen) < 0)
+		cheribsdtest_failure_err("getsockname()");
+	if (socklen != sizeof(sin))
+		cheribsdtest_failure_errx("getsockname() socklen (%u)",
+		    socklen);
+	conn_sock = socket(PF_INET, SOCK_STREAM, 0);
+	if (conn_sock < 0)
+		cheribsdtest_failure_err("socket(PF_INET, SOCK_STREAM)");
+	if (connect(conn_sock, (struct sockaddr *)&sin, sizeof(sin)) < 0)
+		cheribsdtest_failure_err("connect()");
+	accept_sock = accept(listen_sock, NULL, NULL);
+	if (accept_sock < 0)
+		cheribsdtest_failure_err("accept()");
+	fds[0] = conn_sock;
+	fds[1] = accept_sock;
+	close(listen_sock);
+}
+
+/*
+ * Perform a test, sending the tagged pointer over the provided file
+ * descriptor pair and checking various properties on the other side.  Pad the
+ * pointer out to various sizes, triggering (or not) various optimizations.
+ *
+ * Because we want to exercise pipe()'s VM optimisations, we have to use not
+ * just blocking file-descriptor I/O, but also synchronous writes.
+ * Specifically, the sender thread blocks until the receive has received, so
+ * as to avoid needing copy-on-write -- but as a result we can't use the same
+ * thread to send and receive despite buffering.
+ *
+ * To avoid utilising threading, which changes other aspects of the process
+ * execution environment, we fork a process to perform the send, and then do
+ * the receive in the parent process that is running the test.
+ *
+ * Doesn't return.
+ */
+static int pointed_to;
+static void
+ipc_test_tagsend_pointer(int *fds, size_t bufferlen)
+{
+	int * __capability pointer_tosend = &pointed_to;
+	int * __capability pointer_received;
+	void *buffer;
+	ssize_t len;
+	pid_t pid;
+	int status;
+
+	/* Quick self check to make sure we start with a tagged pointer. */
+	if (!cheri_gettag(pointer_tosend))
+		cheribsdtest_failure_errx(
+		    "pointer_tosend untagged before write");
+
+	/* Self check on minimum buffer size. */
+	if (bufferlen < sizeof(pointer_tosend))
+		cheribsdtest_failure_errx("buffer too small");
+
+	/* Self check on alignment of buffer. */
+	if ((((vaddr_t)&buffer) % sizeof(void *)) != 0)
+		cheribsdtest_failure_errx("buffer unaligned");
```

Comment:
If this is not true, the compiler is broken an nothing would work (and the `buffer =` line below would trap anyway for purecap).

Reply:
Fixed by removing 'buffer'.